source/Octopus.Tentacle.Client/Retries/TaskExtensionMethods.cs

```diff
@@ -0,0 +1,27 @@
+﻿using System.Threading.Tasks;
+
+namespace Octopus.Tentacle.Client.Retries
+{
+    static class TaskExtensionMethods
+    {
+        public static void IgnoreUnobservedExceptions(this Task task)
+        {
+            if (task.IsCompleted)
```

Comment:
I wonder if we should do the `if (task.IsCompleted)` check after we set the `ContinueWith`.

Because this is all async, there is a small window of opportunity where the task may complete after we do the 'if' check, but before we register the `ContinueWith`.

Reply:
I don't think there is a race condition. I think it's an optimisation to avoid having to add the `ContinueWith` if not needed

![image](https://github.com/OctopusDeploy/OctopusTentacle/assets/86938706/df33be2d-650e-47a2-96a4-d5926e8a28e0)

```
async Task Main()
{
	// Task not completed so obsevrer with ContinueWith
	var t = Fail();
	t.IgnoreUnobservedExceptions();

	// Task already completed
	var t2 = Fail();
	await Task.Delay(TimeSpan.FromSeconds(2));
	t2.IgnoreUnobservedExceptions();

	// Race condition where task completes after IsCompleted check but before ContinueWith is hooked up
	var t3 = Fail();
	await Task.Delay(TimeSpan.FromSeconds(2));
	t3.IgnoreUnobservedExceptionsWithoutCheckingIsCompleted();

	await Task.Delay(TimeSpan.FromSeconds(10));
}

async Task Fail()
{
	await Task.Delay(TimeSpan.FromSeconds(1));
	throw new Exception();
}

static class TaskExtensionMethods
{
	public static void IgnoreUnobservedExceptions(this Task task)
	{
		if (task.IsCompleted)
		{
			if (task.IsFaulted)
			{
				var observedException = task.Exception;
				"Observed as Task was already Complete".Dump();
			}

			return;
		}

		task.ContinueWith(
			t =>
			{
				var observedException = t.Exception;
				"Observed in ContinueWith".Dump();
			},
			TaskContinuationOptions.OnlyOnFaulted | TaskContinuationOptions.ExecuteSynchronously);
	}

	public static void IgnoreUnobservedExceptionsWithoutCheckingIsCompleted(this Task task)
	{
		if(!task.IsCompleted)
		{
			throw new Exception("Expecting Task to be Completed for this to simulate the race condition");
		}
		
		task.ContinueWith(
			t =>
			{
				var observedException = t.Exception;
				"Observed in ContinueWith".Dump();
			},
			TaskContinuationOptions.OnlyOnFaulted | TaskContinuationOptions.ExecuteSynchronously);
	}
}

// You can define other methods, fields, classes and namespaces here
```

Reply:
Ah, I see. If the task completed after the 'if' but before the `ContinueWith`, then it doesn't matter, because adding the `ContinueWith` will make it run straight away. Like you said, the first part is just an optimisation.